iFrame component

pinussilvestrus opened the pull request:
<!--

Thanks for creating this pull request!

Please make sure to link the issue you are closing as "Closes #issueNr". 
This helps us to understand the context of this PR.

-->

Related to #887

Adds `iFrame` component.
* Specify `url` to embed web content into a form
* Specify title via `label`
* Specify height


- [x] This PR adds a new `form-js` element or visually changes an existing component.
  * => In that case, we need to ensure we follow up on this, e.g. by [creating an issue in Tasklist](https://github.com/camunda/tasklist/issues/new/choose)
  
  
Demo: https://demo-887-iframe-component--camunda-form-playground.netlify.app/

volodymyr-melnykc commented:
@pinussilvestrus The iframe (svg) icon looks too big. I see that it doesn't have inner paddings. Can you please update it to this one?

![iframe icon](https://github.com/bpmn-io/form-js/assets/143816833/9e449ca7-4784-4641-9926-d346babee229)

pinussilvestrus commented on packages/form-js-viewer/src/render/components/Sanitizer.js:
```diff
@@ -45,6 +45,7 @@ const ALLOWED_ATTRIBUTES = [
 
 const ALLOWED_URI_PATTERN = /^(?:(?:(?:f|ht)tps?|mailto|tel|callto|cid|xmpp):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i; // eslint-disable-line no-useless-escape
 const ALLOWED_IMAGE_SRC_PATTERN = /^(https?|data):.*/i; // eslint-disable-line no-useless-escape
+const ALLOWED_IFRAME_SRC_PATTERN = /^(https?):\/\/*/i; // eslint-disable-line no-useless-escape
```
@volodymyr-melnykc @christian-konrad @vsgoulart @Skaiir We could also update this pattern to only allow for `https`, which is recommended anyways.

Wdyt?

vsgoulart replied:
Yeah, I think thatś a good idea

volodymyr-melnykc replied:
Let's also be explicit to the user that we require a `https` URL. 
@pinussilvestrus I updated the copy for the tooltip. Can you please update it in the codebase?

<img width="472" alt="image" src="https://github.com/bpmn-io/form-js/assets/143816833/7e295ac8-039d-4fb2-9127-6595e3d8365b">

pinussilvestrus replied:
Updated it accordingly (preview still building) 👍

pinussilvestrus commented:
@volodymyr-melnykc I updated the icon 👍

pinussilvestrus commented:
TIL: We need to set the `sandbox` attribute explicitly, "empty" is not empty enough.

![image](https://github.com/bpmn-io/form-js/assets/9433996/fdb035cd-21c4-464d-b038-f95dd789e92c)

Thanks to the InfoSec team ❤️

vsgoulart commented on packages/form-js-viewer/src/render/components/form-fields/IFrame.js:
```diff
@@ -0,0 +1,78 @@
+          height={ height }
+          class="fjs-iframe"
+          id={ prefixId(id, formId) }
+          sandbox=""
```
@pinussilvestrus Maybe we should add test for this?

pinussilvestrus replied:
Good catch, added one test case.

pinussilvestrus commented:
> @pinussilvestrus Based on the [latest discussion with InfoSec](https://camunda.slack.com/archives/CKWSYH1S4/p1700129353262419) I updated the tooltip to be more descriptive: https://www.figma.com/file/J78Djf3PgCEAPxFD86floX/Form-Editor-Components?type=design&node-id=465%3A20473&mode=design&t=4Sc34AF6MsKAnHS4-1

Updated it 👍

vsgoulart reviewed (approved):
Only issue I found was the PDF one we discussed via Slack which probably is not fixable from our side

Another small improvement we could make is to display the default height on the properties panel instead of hiding it, what do you think @pinussilvestrus ?

pinussilvestrus commented:
> Another small improvement we could make is to display the default height on the properties panel instead of hiding it, what do you think @pinussilvestrus ?

Yeah that makes sense 👍 I added the default value accordingly.

pinussilvestrus commented:
@esraagamal6 I think you can start testing the demo: https://demo-887-iframe-component--camunda-form-playground.netlify.app/. I don't expect any bigger change for now.

@volodymyr-melnykc @christian-konrad @Skaiir feel free to provide your feedback of course :)

Skaiir commented:
Might be worth having a pattern match or implicitly prefix https:// here: 
![grafik](https://github.com/bpmn-io/form-js/assets/17801113/accdb933-ccb9-4db6-859d-d753f9c3ab63)

pinussilvestrus commented:
@volodymyr-melnykc Do you think the tooltip is enough or should we add a validation message below the static input? Personally I think auto-prefixing is something we should avoid.

Skaiir commented:
This is all take it or leave it opinion btw. The unloaded state feels a little bit technical on the viewer side. I don't know but a user will see this and see iFrame, cool but that might not mean anything to them and it's not explicit visually that something has gone wrong here. 

I'd expect something like, external site couldn't be fetched or something like that. Maybe good as a follow-up.

Wdyt?

![grafik](https://github.com/bpmn-io/form-js/assets/17801113/ef4e4875-88c9-4e26-b40a-221f628ac63b)

Skaiir commented:
> Personally I think auto-prefixing is something we should avoid.

Yeah I agree actually, even if I suggested it :trollface:

pinussilvestrus commented:
> This is all take it or leave it opinion btw. The unloaded state feels a little bit technical on the viewer side. I don't know but a user will see this and see iFrame, cool but that might not mean anything to them and it's not explicit visually that something has gone wrong here.
> 
> I'd expect something like, external site couldn't be fetched or something like that. Maybe good as a follow-up.
> 
> Wdyt?
> 
> ![grafik](https://user-images.githubusercontent.com/17801113/284577711-ef4e4875-88c9-4e26-b40a-221f628ac63b.png)

I think we could be a bit more descriptive if something went wrong. A cheap way would be to place the existing placeholder in the editor and when the source is empty + showing a different text when the URL was not valid after we went through our sanitization.

![image](https://github.com/bpmn-io/form-js/assets/9433996/36f7a6a7-fe36-4dff-a955-a68005a18101)

volodymyr-melnykc commented:
@pinussilvestrus When the URL is not provided, we can show a placeholder in the form preview:
https://www.figma.com/file/J78Djf3PgCEAPxFD86floX/Form-Editor-Components?type=design&node-id=575%3A12427&mode=design&t=0TnKCDbS6b4xH7ex-1

![Form builder - No iFrame URL](https://github.com/bpmn-io/form-js/assets/143816833/388e57ea-0fac-46a2-9084-6270fe95e7b4)

When the provided URL is not "https", it makes sense to show the validation message below the static input (as you suggested).
And if the content can't be loaded, we can show the message in the form preview.
https://www.figma.com/file/J78Djf3PgCEAPxFD86floX/Form-Editor-Components?type=design&node-id=575%3A12815&mode=design&t=0TnKCDbS6b4xH7ex-1

![Form builder - screen](https://github.com/bpmn-io/form-js/assets/143816833/a342008b-35e1-4edf-bc7d-a227da60e31c)

pinussilvestrus commented:
@volodymyr-melnykc I updated it as you suggested 👍

Skaiir commented on packages/form-js-editor/src/features/properties-panel/entries/LabelEntry.js:
```diff
@@ -51,7 +51,12 @@ export default function LabelEntry(props) {
       isEdited: isFeelEntryEdited,
-      isDefaultVisible: (field) => INPUTS.includes(field.type) || field.type === 'button' || field.type === 'group'
+      isDefaultVisible: (field) => (
+        INPUTS.includes(field.type) ||
```
Might make sense to have a parameter for "labeled" at this point, but I don't know. Just a thought for the future, since this conditional is starting to get more and more arbitrary.

Skaiir commented on packages/form-js-editor/src/features/properties-panel/entries/LabelEntry.js:
```diff
@@ -157,4 +162,20 @@ function TimeLabel(props) {
+    return 'Group label';
+  }
+
+  if (type === 'iframe') {
```
Not sure this is super necessary to change the wording here, I presume there were discussions around this?

pinussilvestrus replied:
We kinda misused this one, because technically it's not only a label but also the accessible iFrame title. So we decided to use the existing label for it as it works for both.

Skaiir commented on packages/form-js-editor/src/features/properties-panel/entries/factories/simpleBoolEntryFactory.js:
```diff
@@ -8,6 +8,8 @@ export default function simpleBoolEntryFactory(options) {
     description,
     path,
     props,
+    getValue,
```
Might have to start renaming it to just boolEntryFactory, but I'm not against the change, less boilerplate is always good.

Skaiir commented on packages/form-js-editor/src/features/properties-panel/entries/factories/simpleBoolEntryFactory.js:
```diff
@@ -1,5 +1,5 @@
 import { get } from 'min-dash';
-import { CheckboxEntry, isCheckboxEntryEdited } from '@bpmn-io/properties-panel';
+import { ToggleSwitchEntry, isToggleSwitchEntryEdited } from '@bpmn-io/properties-panel';
```
I've already made this change somewhere else, maybe dynamic list, anyways +1

pinussilvestrus replied:
Let's say I got inspired 😅

Skaiir commented on packages/form-js-editor/test/spec/features/properties-panel/PropertiesPanel.spec.js:
```diff
@@ -31,6 +31,7 @@ import {
 import schema from '../../form.json';
 import defaultValuesSchema from '../../defaultValues.json';
 import redundantValuesSchema from '../../redundantValues.json';
+import iframeSchema from '../../../../../form-js-viewer/test/spec/iframes.json';
```
I'm wondering if generally we should start centralizing our JSON files on a single project, instead of copying across the same JSON for the editor, viewer and playground.

pinussilvestrus replied:
Yeah I thought about a `form-js-testing` package already, could also give good guidance around testing forms in general.

vsgoulart replied:
I had plans on doing that on Tasklist, I would just call it `mocks` instead of testing (unless we want to include some utility functions too)

Skaiir commented on packages/form-js-editor/src/render/components/editor-form-fields/EditorIFrame.js:
```diff
@@ -1,13 +1,19 @@
+  const Icon = iconsByType(field.type);
+
+  return (
+    <div class="fjs-iframe-placeholder">
```
Yeah I actually like this more, less inter-dependency.

pinussilvestrus commented on packages/form-js-carbon-styles/src/carbon-styles.scss:
```diff
@@ -1202,3 +1202,17 @@
+
+// iFrame styles ////////////
+
+.fjs-container {
```
Yeah the thing is, theming in general form-js looks slightly different in comparison to Carbon styles, especially with the layers we have in Tasklist. I tried to use these colors in the base form-js styles as well, but it looks a bit odd with the existing background colors and a theme applied.

pinussilvestrus commented:
QA Testing is done ✅ Merging it now 👍
